teps/0028-task-results-without-results.md

```diff
@@ -0,0 +1,358 @@
+different possible solutions and determine which one fits best to address this issue.
+
+## Motivation
+
+Missing task results should not be fatal. Continue executing `pipeline` when a task result is missing.
+
+### Goals
+
+Identify a set of use cases for `finally` and `continueAfterSkip` without task results and design a common solution
+to address those use cases.
+
+### Non-Goals
+
+This proposal is not offering a solution without any use case.
+
+## Proposal
+
+Enable `pipeline` author to specify a default value for a `task result` in case of a missing result. Pipeline author
+can specify default for the `param` referring to task result if that `param` is declared optional in the `task`.
+
+Generally, task author designs a task with set of inputs in the form of `params` while producing some results in the
+form of `results`. When a `task` is incorporated in a `pipeline`, `pipeline` author has control over these `params`
+and can design it so that the `params` are mapped to `pipeline` params or initialized with the `results` of other
+tasks in a `pipeline`. `params` in a `task` can be declared as required and optional. Required `params` does not allow
+any defaults and have to be initialized at the `pipeline` or `pipelineRun` level.
+
+The proposal is designed on the similar lines, when a `param` is declared `optional` at the `task` level and referring to the
+`task result` of any other task in a `pipeline`, is allowed to have a default at the `pipelineTask` level.
+
+In the following example, `pipelineTask` add declares `results` and its default value. Now, `x` is an `optional` param
+in `multiply` task and referring to a task result of `add`. If `add` fails for some reason, finally task `multiply`
+continues execution with default `sum` which is `0`:
+
+```yaml
+spec:
+  tasks:
+    - name: add
+      taskRef:
+        name: add-task
+      params:
+        - name: first
+          value: $(params.first)
+        - name: second
+          value: $(params.second)
+      results:
+        - name: sum
+          value: 0
+  finally:
+    - name: multiply
+      taskRef:
+        name: multiply
+      params:
+        - name: x
+          value: $(task.add.results.sum)
+```
+
+No default for a `required` task param (referring to a task result) allowed. If a task result is missing, `pipeline`
+skips that `pipelineTask`.
+
+### User Stories
+
+#### Story 1
+
+For a cleanup use case, `finally` task receives the project name, or the configuration name as a `task result` from
+the `acquire` task. If `acquire` task fails to acquire any resources, `cleanup` is skipped.
+
+#### Story 2
+
+Need user story for `finally` task with optional `param` referring to a task result.
+
+#### Story 3
+
+Need user story for a task with `continueAfterSkip` set to `true` with missing task result.
+
+#### Story 4
+
+Need user story for a when expression with missing task result.
```

Comment:
Pipeline that builds a container image and deploys it.
The pipeline accepts an image reference parameter. 
The first task looks up in the registry if the reference exists.
The second task builds the container images. It is guarded via `when` and only runs if the first task reports that the image is missing. 
The third task takes the image reference from a result of the second task and deploys the container.
The pipeline authors sets a the default value for the image reference of the be it's own image reference input param, so that if task2 is skipped, task3 can deploy the image that was passed to the pipeline by the pipelinerun.

```
kind: Pipeline
spec:
  params:
    - name: targetImage
       description: the image to deploy (and build if missing)
  tasks:
    - name: lookup-image
      taskRef:
        name: search-image
      params:
        - name: imageRef
          value: "$(params.targetImage)"
   - name: build-image
      when:
      - input: "$(tasks.lookup-image.results.found)"
        operator: in
        values: ["False"]
      taskRef:
        name: build-image
      params:
        - name: imageRef
          value: "$(params.targetImage)"
   - name: deploy-image
      taskRef:
        name: deploy-image
      params:
        - name: imageRef
           value: "$(tasks.build-image.results.builtImage)"
           default: "$(params.targetImage)"
```

Reply:
thank you for sample use case, have included in the TEP